storm/src/main/storm/mesos/resources/RangeResource.java

```diff
@@ -0,0 +1,214 @@
+    if (isAvailable(rangeResourceEntry)) {
+      List<ReservationType> reservationTypeList = Arrays.asList(ReservationType.values());
+      Collections.sort(reservationTypeList, reservationTypeCompartor);
+      return removeAndGet(reservationTypeList, rangeResourceEntry);
+    }
+
+    String message = String.format("ResourceType '%s' is not available. Requested value: %s Available: %s",
+                                   resourceType, rangeResourceEntry, toString());
+    throw new ResourceNotAvailabeException(message);
+  }
+
+  public String toString(List<RangeResourceEntry> ranges) {
+    List<String> rangeStrings = new ArrayList<>();
+    for (RangeResourceEntry range : ranges) {
+      String beginStr = String.valueOf(range.getBegin());
+      String endStr = String.valueOf(range.getEnd());
+      /*
+      * A Range representing a single number still has both Range.begin
+      * and Range.end populated, but they are set to the same value.
+      * In that case we just return "N" instead of "N-N".
+      */
+      if (range.getBegin() == range.getEnd()) {
+        rangeStrings.add(beginStr);
+      } else {
+        rangeStrings.add(String.format("%s-%s", beginStr, endStr));
+      }
+    }
+    return String.format("[%s]", StringUtils.join(rangeStrings, ","));
+  }
+
+  public String toString(ReservationType reservationType) {
+    return String.format("%s : %s", reservationType, toString(availableResourcesByReservationType.get(reservationType)));
+  }
+
+  public String toString() {
+    List<RangeResourceEntry> resourceRanges = new ArrayList<>();
+    for (Map.Entry<ReservationType, List<RangeResourceEntry>> entry : availableResourcesByReservationType.entrySet()) {
+      resourceRanges.addAll(entry.getValue());
+    }
+    return toString(resourceRanges);
+  }
+
+  private List<ResourceEntry> removeAndGet(Collection<ReservationType> reservationTypes, RangeResourceEntry rangeResourceEntry) {
+    List<ResourceEntry> removedResources = new ArrayList<>();
+    Long beginValue = rangeResourceEntry.getBegin();
+    Long endValue = rangeResourceEntry.getEnd();
+
+    for (ReservationType reservationType : reservationTypes) {
+      List<RangeResourceEntry> rangeResourceEntryList = availableResourcesByReservationType.get(reservationType);
+      for (int i = 0; i < rangeResourceEntryList.size(); i++) {
+        RangeResourceEntry tmp = rangeResourceEntryList.get(i);
+        if (beginValue >= tmp.getBegin() && endValue <= tmp.getEnd()) {
+          rangeResourceEntryList.remove(i);
+          // We already removed the entry. So when beginValue == endValue,
+          // we dont have to add a new entry
+          if (tmp.getBegin().equals(tmp.getEnd()) || (tmp.getBegin().equals(beginValue) && tmp.getEnd().equals(endValue))) {
+            removedResources.add(tmp);
+            return removedResources;
+          }
+
+          if (beginValue > tmp.getBegin() && tmp.getEnd().equals(endValue)) {
```

Comment:
I've been thinking about this, and we do a lot of repeated things in these blocks, so what if we refactored the logic inside the (first `if` to this) this:
```
availableRanges.remove(i);
// If this is exactly the ranges we were looking for, then we can use them
if (availableRange.getBegin().equals(availableRange.getEnd()) || (availableRange.getBegin().equals(desiredBegin) && availableRange.getEnd().equals(desiredEnd))) {
    removedResources.add(availableRange);
    return removedResources;
}
// Salvage resources before the beginning of the requested range
if (availableRange.getBegin() < desiredBegin) {
    availableRanges.add(new RangeResourceEntry(reservationType, availableRange.getBegin(), desiredBegin() - 1);
}
// Salvage resources after the end of the requested range
if (availableRange.getEnd() > desiredEnd) {
    availableRanges.add(new RangeResourceEntry(reservationType, desiredEnd + 1, availableRange.getEnd()));
}
// Now that we've salvaged all available resources, add the resources for the specifically requested range
removeResources.add(new RangeResourceEntry(reservationType, desiredBegin, desiredEnd));
```